### ISSUE TITLE
Spot wrong usage of bitwise and operator

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust-clippy#1594]
I'd like Clippy to warn in this expression suggesting that this is probably an error, and that the programmer perhaps meant to use "&&" instead:

```
fn foo2(x1: i32, x2: i32) {
    if (x1 < x2) & (x1 != 0) {}
}
```

**Comment by mcarton** [comment]
/me disappointed
/me expected that to be a type-error in rustc

**Comment by leonardo-m** [comment]
In my opinion there's a small number of Clippy lints that should be Rustc errors (or warnings).

**Comment by mcarton** [comment]
What I meant is I did not expect `&` to even exist on booleans.

**Comment by oli-obk** [comment]
Actually there are valid use cases for this, because the regular bool ops  short circuit but in that case the lint can be disabled, since these cases are rare.

**Comment by clarcharr** [comment]
Perhaps this could only apply to constant operations and functions that are known to have no side effects? That way those cases don't require disabling the lint.

**Comment by mcarton** [comment]
@oli-obk I know but this is rare enough that I would have expected to have to bind the expressions (and of course add a comment) rather that be allowed to use `&` on booleans. Oh well, it's too late anyway.

**Comment by mcarton** [comment]
Wrong button (morning 😅).

**Comment by leonardo-m** [comment]
> Wrong button (morning

It's a common mistake, so I've sent a "bug report" to GitHub about the design of those two buttons.

**Comment by leonardo-m** [comment]
Is the example in #1594 not spotted by this lint?

```rust
fn foo2(x1: i32, x2: i32) {
    if (x1 < x2) & (x1 != 0) {}
}
```
